Sources/NIOCore/IPAddresses.swift

```diff
@@ -0,0 +1,447 @@
+                                    + UInt32(self.address.bytes.2) << 8
+                                    + UInt32(self.address.bytes.3))
+            }
+        }
+        
+        /// Creates a new `IPv4Address`.
+        ///
+        /// - parameters:
+        ///   - address: Bytes that hold the IPv4 address.
+        fileprivate init(address: IPv4Bytes) {
+            self.address = address
+        }
+        
+        /// Creates a new `IPv4Address`.
+        ///
+        /// - parameters:
+        ///   - packedBytes: Collection of UInt8 that holds the address.
+        fileprivate init<Bytes: Collection>(packedBytes bytes: Bytes) where Bytes.Element == UInt8, Bytes.Index == Int {
+            self = .init(address: .init((
+                bytes[0], bytes[1], bytes[2], bytes[3]
+            )))
+        }
+        
+        /// Creates a new `IPv4Address`.
+        ///
+        /// - parameters:
+        ///   - string: String representation of an IPv4 address.
+        fileprivate init(string: String) throws {
+            var bytes: [UInt8] = [0,0,0,0]
+            var idx: Int = 0
+            
+            for char in string {
+                if char == "." {
+                    idx += 1
+                } else if let number = char.wholeNumberValue {
+                    if idx > 3 || bytes[idx] > 25 || (255 - number < (bytes[idx] * 10)) {
+                        throw IPAddressError.failedToParseIPString(string)
+                    }
+                    bytes[idx] = bytes[idx] * 10 + UInt8(number)
+                } else {
+                    throw IPAddressError.failedToParseIPString(string)
+                }
+            }
+            if idx != 3 {
+                throw IPAddressError.failedToParseIPString(string)
+            }
+            self = .init(packedBytes: bytes)
+        }
+    }
+    
+    /// A single IPv6 address for `IPAddress`
+    public struct IPv6Address {
+        /// The bytes storing the address of the IPv6 address.
+        let address: IPv6Bytes
+        
+        /// Get the `IPv6Address` as a string.
+        var ipAddress: String {
+            stride(from: 0, to: 15, by: 2).lazy.map({ idx in
+                let hexValues = [
+                    self.address[idx] >> 4,
+                    self.address[idx] & 0x0F,
+                    self.address[idx + 1] >> 4,
+                    self.address[idx + 1] & 0x0F
+                ]
+                
+                var removeLeadingZeros = 0
+                if hexValues[0] + hexValues[1] + hexValues[2] == 0 {
+                    removeLeadingZeros = 3
+                } else if hexValues[0] + hexValues[1] == 0 {
+                    removeLeadingZeros = 2
+                } else if hexValues[0] == 0 {
+                    removeLeadingZeros = 1
+                }
+                
+                return String(hexValues[removeLeadingZeros...].lazy.map {$0.toHex()})
+            }).joined(separator: ":")
```

Comment:
This doesn't correctly handle runs of leading zeros in the middle, where it will emit those as `0:0:0:0` instead of as `::`.